Esti: Refactor gateway tests

N-o-Z opened the pull request:
## Change Description

### Background

Part of the effort to stabilize and fix esti tests.
Update all tests to use S3 client instead of minio client with the exception of one test which needs minio for the V2 signing scenario (which is no longer supported in AWS SDK)

arielshaqed commented:
Drive-by review: this PR will effectively remove support for sigv2 from lakeFS.  While I _really_ understand the pain, I'm not sure we can do that without checking how or users really use it.

The description of this PR f states that we want to stop using the MinIO client in Esti, in order to "stabilize and fix" it.  How will this help? It's just a client. Its use in Esti is not broken.  At best it means we won't need to configure and use two clients.

Why do we use it? **Because** it's a way to test sigv2. As you rightly state, the AWS Go _clients_ don't support this signing method - this has been true for about the entire lifetime of lakeFS. However we have seen sigv2 used with lakeFS. Also note that V2 has certain insecurity features, such as a longer lifetime for presigned URLs - detailed on [this blog](https://aws.amazon.com/blogs/aws/amazon-s3-update-sigv2-deprecation-period-extended-modified/). That blog mentions a deprecation date in 2019 (`#not_a_typo`)... and yet here we are.  It's 2025, and AWS still support it on the server side AFAIK, at least in some regions.

arielshaqed commented:
# 🧟 

For fun, [here's a fix](https://github.com/aws/aws-sdk-net/issues/3167) to an official AWS SDK from just over a year ago.

arielshaqed commented:
> # 🧟
> For fun, [here's a fix](https://github.com/aws/aws-sdk-net/issues/3167) to an official AWS SDK from just over a year ago.

[Here's](https://forum.lakefs.io/t/14168102/hi-again-guys-really-like-lakefs-i-have-a-problem-regarding-#d26988d3-4dae-45ac-8f24-52a1ef89f31d) our @rmoff using V2 while helping some user, 2 years ago. The point is that it easily gets used, not that it was configured correctly there.

nopcoder commented on esti/s3_gateway_test.go:
```diff
@@ -143,21 +174,23 @@ func TestS3UploadAndDownload(t *testing.T) {
 						}
 
-						download, err := client.GetObject(ctx, repo, o.Path, minio.GetObjectOptions{})
+						download, err := clt.GetObject(ctx, repo, o.Path, minio.GetObjectOptions{})
```
based on the other code here it looks like we don't close the response body for this call. unless there is no need to close the response body for minio client GetObject and we need to remove the close from other places.

nopcoder reviewed (changes requested):
Minor comments and question

N-o-Z commented:
@nopcoder @arielshaqed 
I've scratched the branch and took a different route to the changes.
Modified the s3 client to take into consideration secure endpoint which resolved most of the issues in CI.
Also modified how we generate the random object path according to AWS s3 naming requirements

nopcoder commented on pkg/testutil/random.go:
```diff
@@ -65,3 +65,14 @@ func NewRandomReader(rand *rand.Rand, size int64) io.Reader {
 	return nanoid.MustGenerate(chars, charsSize)
 }
+
+func RandomS3Path(rand *rand.Rand, size int) string {
+	// List of valid chars per https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html
+	var pathRunes = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!-_.*'()&$@=;/:+,?")
+
+	b := make([]rune, size)
+	for i := range b {
+		b[i] = pathRunes[rand.Intn(len(pathRunes))]
+	}
+	return string(b)
+}
```
- Can we make the charset constant?
- Can we use `gonanoid.Generate` to generate the random key

N-o-Z commented on pkg/testutil/setup.go:
```diff
@@ -139,9 +139,14 @@ func SetupTestingEnv(params *SetupTestingEnvParams) (logging.Logger, apigen.Clie
 
-func SetupTestS3Client(endpoint, key, secret string, forcePathStyle bool) (*s3.Client, error) {
+func SetupTestS3Client(endpoint, key, secret string, forcePathStyle bool, opts ...func(*s3.Options)) (*s3.Client, error) {
+	endpointSecure := viper.GetBool("s3_endpoint_secure")
 	if !strings.HasPrefix(endpoint, "http") {
-		endpoint = "http://" + endpoint
+		if endpointSecure {
+			endpoint = "https://" + endpoint
+		} else {
+			endpoint = "http://" + endpoint
+		}
```
This is the env var name used also for the minio client, the package level client and in the action files and docker compose files.
I'd prefer not to make the changes as part of this PR

nopcoder commented on pkg/testutil/setup.go:
```diff
@@ -150,10 +155,10 @@ func SetupTestS3Client(endpoint, key, secret string, forcePathStyle bool) (*s3.C
 		return nil, err
 	}
-	svc := s3.NewFromConfig(cfg, func(options *s3.Options) {
+	svc := s3.NewFromConfig(cfg, append(opts, func(options *s3.Options) {
```
Just wanted to be clear you will not be able to override them using opts

nopcoder reviewed (approved):
lgtm
